# Part A: Dataset Understanding 
#### Q1. Load the dataset and display the first five records.

In [2]:
import pandas as pd
# Load dataset
df = pd.read_csv("netflixdataset.csv")

# Display first five rows
df.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


#### Q2. Determine the number of rows and columns in the dataset

In [4]:
# Number of rows and columns
print("Rows and Columns:", df.shape)

Rows and Columns: (750, 10)


#### Q3. Display all column names

In [5]:
# Display column names
print(df.columns)

Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


#### Q4. Identify numerical and categorical features

In [6]:
# Identify numerical and categorical columns

numerical_features = df.select_dtypes(include=['int64', 'float64']).columns
categorical_features = df.select_dtypes(include=['object']).columns

print("Numerical Features:")
print(numerical_features)

print("\nCategorical Features:")
print(categorical_features)

Numerical Features:
Index(['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks',
       'MonthlySpend'],
      dtype='object')

Categorical Features:
Index(['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed'], dtype='object')


#### Q5. Check whether the dataset contains missing values

In [7]:
# Check missing values
print(df.isnull().sum())

UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64


---
# Part B: Exploratory Data Analysis (EDA)

#### Q6. Calculate the average age of users

In [8]:
# Average age of users
average_age = df['Age'].mean()

print("Average Age:", average_age)

Average Age: 41.824


#### Q7. Determine the average watch hours per week

In [9]:
# Average watch hours per week
average_watch_hours = df['WatchHoursPerWeek'].mean()

print("Average Watch Hours Per Week:", average_watch_hours)

Average Watch Hours Per Week: 14.236


#### Q8. Find the average monthly spending of users

In [11]:
# Average monthly spending
average_spending = df['MonthlySpend'].mean()

print("Average Monthly Spending:", average_spending)

Average Monthly Spending: 689.9053333333334


#### Q9. Count the number of users in each subscription category

In [12]:
# Count users in each subscription category
subscription_count = df['SubscriptionType'].value_counts()

print(subscription_count)

SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


#### Q10. Determine the percentage of users who renewed their subscriptions

In [13]:
# Percentage of users who renewed subscription

renewal_percentage = (df['SubscriptionRenewed'].value_counts(normalize=True) * 100)

print(renewal_percentage)

SubscriptionRenewed
No     53.733333
Yes    46.266667
Name: proportion, dtype: float64


---
# Part C — Data Preparation
#### Q11. Convert categorical features into numerical form

In [14]:
from sklearn.preprocessing import LabelEncoder

# Create LabelEncoder object
le = LabelEncoder()

# Convert categorical columns into numerical form
for column in categorical_features:
    df[column] = le.fit_transform(df[column])

# Display updated dataset
df.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,0,0,23,5,1,13,353,0
1,1002,55,1,0,9,4,2,14,317,1
2,1003,49,1,0,8,3,1,16,309,0
3,1004,39,0,1,19,5,2,45,833,1
4,1005,38,0,1,23,5,5,24,804,1


#### Q12. Define the feature set (X) and target variable (y)

In [18]:
# Define features and target variable

X = df.drop('SubscriptionRenewed', axis=1)
y = df['SubscriptionRenewed']

print("Features:")
print(X.head())

print("\nTarget:")
print(y.head())


Features:
   UserID  Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22       0                 0                 23            5   
1    1002   55       1                 0                  9            4   
2    1003   49       1                 0                  8            3   
3    1004   39       0                 1                 19            5   
4    1005   38       0                 1                 23            5   

   FavoriteGenre  AdClicks  MonthlySpend  
0              1        13           353  
1              2        14           317  
2              1        16           309  
3              2        45           833  
4              5        24           804  

Target:
0    0
1    1
2    0
3    1
4    1
Name: SubscriptionRenewed, dtype: int64


#### Q13. Split the dataset into training and testing sets

In [19]:
from sklearn.model_selection import train_test_split

# Split dataset into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

print("Training Data Shape:", X_train.shape)
print("Testing Data Shape:", X_test.shape)

Training Data Shape: (600, 9)
Testing Data Shape: (150, 9)


---
# Part D — Decision Tree Classification
#### Q14. Train a Decision Tree model to predict subscription renewal

In [20]:
from sklearn.tree import DecisionTreeClassifier

# Create Decision Tree model
dt_model = DecisionTreeClassifier(random_state=42)

# Train the model
dt_model.fit(X_train, y_train)

print("Decision Tree model trained successfully!")

Decision Tree model trained successfully!


#### Q15. Evaluate the model using accuracy

In [21]:
from sklearn.metrics import accuracy_score

# Make predictions
y_pred_dt = dt_model.predict(X_test)

# Calculate accuracy
dt_accuracy = accuracy_score(y_test, y_pred_dt)

print("Decision Tree Accuracy:", dt_accuracy)

Decision Tree Accuracy: 0.5666666666666667


#### Q16. Generate and interpret the confusion matrix

In [22]:
from sklearn.metrics import confusion_matrix

# Generate confusion matrix
cm = confusion_matrix(y_test, y_pred_dt)

print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[50 32]
 [33 35]]


---
# Part E — K-Nearest Neighbors (KNN)
#### Q17. Train a KNN classifier with K = 5

In [23]:
from sklearn.neighbors import KNeighborsClassifier

# Create KNN model
knn_model = KNeighborsClassifier(n_neighbors=5)

# Train the model
knn_model.fit(X_train, y_train)

# Make predictions
y_pred_knn = knn_model.predict(X_test)

print("KNN model trained successfully!")

KNN model trained successfully!


#### Q18. Compare the accuracy of KNN with the Decision Tree model

In [24]:
# Calculate KNN accuracy
knn_accuracy = accuracy_score(y_test, y_pred_knn)

print("Decision Tree Accuracy:", dt_accuracy)
print("KNN Accuracy:", knn_accuracy)

Decision Tree Accuracy: 0.5666666666666667
KNN Accuracy: 0.6


---
# Part F — Linear Regression
#### Q19. Train a Linear Regression model to predict monthly spending

In [26]:
from sklearn.linear_model import LinearRegression

# Define features and target for regression
X_reg = df.drop('MonthlySpend', axis=1)
y_reg = df['MonthlySpend']

# Split data
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg, y_reg,
    test_size=0.2,
    random_state=42
)

# Create Linear Regression model
lr_model = LinearRegression()

# Train model
lr_model.fit(X_train_reg, y_train_reg)

print("Linear Regression model trained successfully!")

Linear Regression model trained successfully!


#### Q20. Predict the monthly spending for a new user and interpret the result

In [27]:
# Predict monthly spending for a new user

new_user = [X_test_reg.iloc[0]]

predicted_spending = lr_model.predict(new_user)

print("Predicted Monthly Spending:", predicted_spending[0])

Predicted Monthly Spending: 838.5871865891318


C:\Users\asus\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2749: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


---

# Business Reflection Questions

## 1. Which factors appear to influence subscription renewal the most?

The factors that likely influence subscription renewal are:

- **WatchHoursPerWeek** → users who spend more time on the platform are more likely to renew.
- **MonthlySpend** → users spending more money may show higher engagement and loyalty.
- **SubscriptionType** → Premium and VIP users may behave differently compared to Basic users.
- **AdClicks** and **FavoriteGenre** may also affect engagement and renewal behavior.

These features help the model identify patterns related to customer retention.

---

## 2. Why is subscription renewal a classification problem?

Subscription renewal is a classification problem because the target variable **SubscriptionRenewed** has categorical outputs:

- Yes
- No

The model predicts which category/class a user belongs to.

---

## 3. Why is monthly spending a regression problem?

Monthly spending is a regression problem because:

- **MonthlySpend** is a continuous numerical value.
- The model predicts an exact amount (for example, 838.59).

Regression algorithms are used when the output is numeric rather than categorical.

---

## 4. Which algorithm performed better for renewal prediction?

The **KNN model performed better** than the Decision Tree model.

### Model Accuracies:
- Decision Tree Accuracy = **56.67%**
- KNN Accuracy = **60%**

Since KNN achieved higher accuracy, it provided slightly better prediction performance for subscription renewal.

---

## 5. How could the platform use these predictions to improve customer retention?

The platform can use these predictions to:

- identify users who are less likely to renew,
- provide personalized offers and discounts,
- recommend content based on favorite genres,
- improve engagement for low-activity users,
- target users with retention campaigns before subscription expiry.

This can help increase customer satisfaction and improve subscription renewal rates.

---